torchtune/models/gemma2/_attention.py

```diff
@@ -0,0 +1,305 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import logging
+from typing import Optional
+
+import torch
+import torch.nn.functional as F
+from torch import nn
+from torchtune.modules.attention_utils import _MaskType
+from torchtune.modules.kv_cache import KVCache
+
+
+logger = logging.getLogger(__name__)
+
+
+class Gemma2Attention(nn.Module):
```

Comment:
Since we support flex attention, which support soft capping, would it make sense to just force gemma2 users to use flex attention instead of implementing this module?

Reply:
Flex Attention is only supported on A100 or better, right? I don't think we can make the assumption that our users will have that.

Reply:
Hello everyone,

I just pushed a new commit which includes all changes discussed with @ebsmothers.
I also implemented a flex attention version but I could not make it work properly.


The default implementation (not using FlexAttention) seems to be working (I only launched the single lora pipeline, please see the attached logs [log_gemma2-2b-single-lora_1729498141.txt](https://github.com/user-attachments/files/17473404/log_gemma2-2b-single-lora_1729498141.txt)).


I would appreciate some help on the FlexAttention implementation. Here is why I am struggling.

If I run the following code on my A6000 GPU with torch 2.5:
```
import torch

from torch.nn.attention.flex_attention import (
    create_block_mask,
    flex_attention)


WINDOW_SIZE=None #None
CAPPING=50.
SCALE=12.


def get_gemma2_flex_score_mask(sliding_window_size, softcapping, query_pre_attn_scalar):
    
    def sliding_window_causal_mask(b, h, q_idx, kv_idx):
        """Causal mask and sliding window as proposed here:
        https://github.com/pytorch-labs/attention-gym/blob/main/examples/flex_attn.ipynb
        """
        causal_mask = q_idx >= kv_idx
        if sliding_window_size is None:
            # if no sliding window return causal mask
            return causal_mask
        else:
            windowed_mask = q_idx - kv_idx <= sliding_window_size

            return causal_mask & windowed_mask
    
    def soft_capping_with_scaling(score, b, h, q_idx, kv_idx):
        if query_pre_attn_scalar is None:
            # usual scaling included in FlexAttention
            score = score / softcapping
            score = torch.tanh(score) #tanh_approx(score)
            return score * softcapping
        else:
            score = score / softcapping * query_pre_attn_scalar**-0.5
            score = torch.tanh(score) #tanh_approx(score)
            return score * softcapping
    
    return sliding_window_causal_mask, soft_capping_with_scaling

# Compile the flex_attention function
flex_attention = torch.compile(flex_attention, dynamic=False)

B=4
H=8
S=117
D=256 #256

mask_mod, score_mod = get_gemma2_flex_score_mask(WINDOW_SIZE, CAPPING, SCALE)

query = torch.randn(
        B, H, S, D, device="cuda", dtype=torch.float16, requires_grad=True
    )
key = torch.randn(
    B, H, S, D, device="cuda", dtype=torch.float16, requires_grad=True
)
value = torch.randn(
    B, H, S, D, device="cuda", dtype=torch.float16, requires_grad=True
)
gradOut = torch.randn(B, H, S, D, device="cuda", dtype=torch.float16)


block_mask = create_block_mask(mask_mod=mask_mod,
                                   B=1,
                                   H=1,
                                   Q_LEN=S,
                                   KV_LEN=S,
                                   device=query.device)

out = flex_attention(
    query, key, value, score_mod=score_mod, block_mask=block_mask
)
print(out.shape)
```

The code runs fine if I don't compile the flex attention by commenting `flex_attention = torch.compile(flex_attention, dynamic=False)` but it raises this error otherwise:
`BackendCompilerFailed: backend='inductor' raised:
OutOfResources: out of resource: shared memory, Required: 114688, Hardware limit: 101376. Reducing block sizes or `num_stages` may help.`

So I disabled compilation and the code seems to be running but very very slowly (48s per iteration vs 1-2s on non flex implementation).

Maybe you could help me understand what is going on ?

Reply:
Tagging @RdoubleA and @felipemello1 for their thoughts. 

Just checking: which size Gemma-2 model are you testing with?

Reply:
The logs I shared are from gemma2 2B, the code snippet is independent of the gemma architecture it's just a toy example.
I am currently running the qlora single device pipeline with 9B (without flex attention), I'll share the logs tomorrow (I'll push the changes to recipe as there are typos on the output path etc).

Reply:
@Optimox just took a look at this. Your snippet here runs for me on an A100, so maybe it is an issue with compile on your hardware. I will investigate more, but in the meantime also tagging @SalmanMohammadi who I believe saw the same error previously.

Reply:
[Possibly relevant suggestion](https://github.com/pytorch/pytorch/issues/133254#issuecomment-2408710459)

Reply:
I already tried to change the kernels but I still can't compile.

Reply:
To add more information on this, if I set the kernel options as discussed [here](https://github.com/pytorch/pytorch/issues/133254#issuecomment-2408710459) I get a different error `LoweringException` and 
```
BackendCompilerFailed: backend='inductor' raised:
LoweringException: AssertionError: 
  target: flex_attention
  args[0]: TensorBox(StorageBox(
    InputBuffer(name='primals_1', layout=FixedLayout('cuda', torch.float16, size=[4, 8, 117, 256], stride=[239616, 29952, 256, 1]))
  ))
  args[1]: TensorBox(StorageBox(
    InputBuffer(name='primals_2', layout=FixedLayout('cuda', torch.float16, size=[4, 8, 117, 256], stride=[239616, 29952, 256, 1]))
  ))
  args[2]: TensorBox(StorageBox(
    InputBuffer(name='primals_3', layout=FixedLayout('cuda', torch.float16, size=[4, 8, 117, 256], stride=[239616, 29952, 256, 1]))
  ))
  args[3]: Subgraph(name='sdpa_score0', graph_module=<lambda>(), graph=None)
  args[4]: (TensorBox(StorageBox(
    InputBuffer(name='primals_4', layout=FixedLayout('cuda', torch.int32, size=[1, 1, 1], stride=[1, 1, 1]))
  )), TensorBox(StorageBox(
    InputBuffer(name='primals_5', layout=FixedLayout('cuda', torch.int32, size=[1, 1, 1, 1], stride=[1, 1, 1, 1]))
  )), TensorBox(StorageBox(
    InputBuffer(name='primals_6', layout=FixedLayout('cuda', torch.int32, size=[1, 1, 1], stride=[1, 1, 1]))
  )), TensorBox(StorageBox(
    InputBuffer(name='primals_7', layout=FixedLayout('cuda', torch.int32, size=[1, 1, 1, 1], stride=[1, 1, 1, 1]))
  )), TensorBox(StorageBox(
    InputBuffer(name='primals_8', layout=FixedLayout('cuda', torch.int32, size=[1, 1, 1], stride=[1, 1, 1]))
  )), TensorBox(StorageBox(
    InputBuffer(name='primals_9', layout=FixedLayout('cuda', torch.int32, size=[1, 1, 1, 1], stride=[2, 2, 1, 2]))
  )), TensorBox(StorageBox(
    InputBuffer(name='primals_10', layout=FixedLayout('cuda', torch.int32, size=[1, 1, 1], stride=[1, 1, 1]))
  )), TensorBox(StorageBox(
    InputBuffer(name='primals_11', layout=FixedLayout('cuda', torch.int32, size=[1, 1, 1, 1], stride=[2, 2, 1, 2]))
  )), 128, 128, Subgraph(name='sdpa_mask0', graph_module=<lambda>(), graph=None))
  args[5]: 0.0625
  args[6]: {'BLOCK_M': 64, 'BLOCK_N': 64, 'BLOCK_M1': 32, 'BLOCK_N1': 64, 'BLOCK_M2': 64, 'BLOCK_N2': 32, 'ROWS_GUARANTEED_SAFE': False, 'PRESCALE_QK': False, 'OUTPUT_LOGSUMEXP': True}
  args[7]: ()
  args[8]: ()
```

If I comment the compilation and use the new kernel options the code runs fine.

Reply:
Hi @Optimox for the forward pass the only relevant options are `BLOCK_M` and `BLOCK_N` IIRC. If you set your sequence length to a multiple of 128 (`S=128`, `S=8192`, etc.) and `kernel_options = {"BLOCK_M": 32, "BLOCK_N": 32}` the code you wrote should work on 3090s/4090s/etc. I think the issue is related to the way the default kernel options are chosen right now for flex [(e.g. for fwd pass)](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/kernel/flex_attention.py#L594), which will treat A6000/3090/4090 as an A100, even though max shared memory is different:

```python
import torch
import triton

def get_shared_memory_config():
    device = torch.cuda.current_device()
    # Get current shared memory config
    current_shared = triton.runtime.driver.active.utils.get_device_properties(device)[
        "max_shared_mem"
    ]
    return current_shared
    
# 4090 = a6000 = 3090: 101376    
# A100: 166912
```
Hopefully this is improved in 2.5.1.

Reply:
Hello @mirceamironenco, thanks for your help, with S a multiple of 128 and `kernel_options = {"BLOCK_M": 32, "BLOCK_N": 32}` the compilation works indeed! Would this configuration also work for a100 ?
Could you explain what `max_share_mem` is here and why we should pick 32 ?

However, in my code snippet I used S=117 because this was the real size of my first batch when lauchning the training recipe. This means I should change the datasets and dataloaders to be sure that sequence length are padded to a multiple of 128.

Reply:
- For now (pt 2.5.0), FlexAttention requires that all sequence lengths be a multiple of 128. IIUC this is being worked on.
- A100/H100 should work without any kernel_options, I think for (head_dim=256, dtype=bfloat16) it will use (32, 64) and (64, 32) respectively. Normally you could set `torch.compile(..., mode="max-autotune")` to have it try out a few options, but this is broken for flex_attention on 2.5.0. Unfortunately, for head_dim=256 3090s/4090s/etc will require kernel_options. Maybe this can be an environment variable until the defaults improve in pt.
- `max_shared_mem` is cuda shared memory which is used by the triton kernel generated by `torch.compile`. The amount of memory used by the kernel is dependent on e.g. dtype, block_N, block_M, num_stages. I just choose the next smaller multiple of 2 (64 -> 32).

Reply:
cc @drisspg @yanboliang in regards to [this comment above](https://github.com/pytorch/torchtune/pull/1835#discussion_r1815058279), is it possible to update the default kernel options in flex for better support of A6000/3090/4090s?

Reply:
With the given kernel options, flex attention can be compiled and the code runs (9b lora single device training). However, the code is terribly slow (29 tokens per second) and the loss turns to nan after one batch:
```
Step 1 | loss:87.6104507446289 lr:2.0000000000000003e-06 tokens_per_second_per_gpu:21.504354449638843 
Step 2 | loss:nan lr:4.000000000000001e-06 tokens_per_second_per_gpu:29.156293709460176 
```

I don't understand what I am doing wrong, the only obvious optimisation I see is to create one block mask for every layer while I am currently recreating the same block mask for every layer (line 593 in `gemma2/_attention.py`). Nevertheless, I do not think that this is the current bottleneck.

Wouldn't it be better to go with the simpler implementation for now and switch to FlexAttention when it will work on more GPUs? or at least leave the choice of computation to the final user and default to the classical implementation ?

Reply:
Yeah I see what you're saying.. I repro'd this on my end too so it is not a function of any custom kernel configs you're using. Let me look into this a bit more but in the meantime it seems like we shouldn't enable the flex version until we figure this out

Reply:
I have updated the code to keep the flex attention implementation but disable it for now, until we have found a solution.

Reply:
Hey sorry just catching up:

So 2.5 should not require multiple of 128 for sequence length. It is unfortunately pretty common for consumer gpus to hit the SharedMemory issue. I have a pr: https://github.com/pytorch/pytorch/pull/137959 to drop default block sizes but still need to debug the failing test.

For being slow, it is expected that the tanh instruction is very slow compared to the inline assembly variant: https://github.com/pytorch-labs/attention-gym/blob/36f8bd5ded5b3469f7892099590bb2405cc8f744/attn_gym/mods/softcapping.py#L92.


It is actually quite hard generically to know what what block sizes should be used since the amount of shared memory depends on the captured buffers. I am working on a better solution but that is going to take some time unfortunately

Reply:
Have you all figured out a solution to the out of resource: shared memory? Seems like any large hidden dim >=128 causes issues for me.